pgmpy/models/DynamicBayesianNetwork.py

```diff
@@ -533,6 +534,215 @@ def add_cpds(self, *cpds):
+        - Variables in subsequent time slices (t>0) use conditional probabilities
+          given their parents from the same and previous time slices
+
+        Parameters
+        ----------
+        data: pd.DataFrame instance
+            The dataset against which to score the model.
+            Each column should represent a variable in the form of (node_name, time_slice).
+
+        show_progress : bool, default=False
+            Whether to display a progress bar during computation.
+
+        Returns
+        -------
+        float
+            The total log-likelihood score for the entire dataset. This is the
+            sum of log-likelihood scores for all individual data points.
+            More negative values indicate worse fit to the data.
+
+        Raises
+        ------
+        ValueError
+            If data is not a pandas DataFrame, contains missing values,
+            has incorrect column structure, or contains values not present
+            in the model's CPD state names.
+
+        Examples
+        --------
+        Examples
+        --------
+        >>> import numpy as np
+        >>> import pandas as pd
+        >>> from pgmpy.models import DynamicBayesianNetwork as DBN
+        >>>
+        >>> # Create a simple DBN with temporal dependencies
+        >>> model = DBN(
+        ...     [(("A", 0), ("B", 0)), (("A", 0), ("A", 1)), (("B", 0), ("B", 1))]
+        ... )
+        >>>
+        >>> # Generate sample data with 4 variables across 3 time slices
+        >>> data = np.random.randint(low=0, high=2, size=(100, 6))
+        >>> colnames = [("A", 0), ("B", 0), ("A", 1), ("B", 1), ("A", 2), ("B", 2)]
+        >>> df = pd.DataFrame(data, columns=colnames)
+        >>>
+        >>> # Fit model and compute log-likelihood
+        >>> model.fit(df)
+        >>> ll_score = model.log_likelihood(df)
+        >>> print(f"Log-likelihood: {ll_score:.2f}")
+        Log-likelihood: -138.62
+        """
+
+        # Input validation
+        if not isinstance(data, pd.DataFrame):
+            raise ValueError(
+                f"Data must be a pandas.DataFrame instance. Got: {type(data)}"
+            )
+
+        if min(data.columns, key=lambda x: x[1])[1] != 0:
+            raise ValueError("Data column names must start from time slice 0.")
+
+        if set(self.nodes()) - set(data.columns):
+            raise ValueError(
+                f"Missing columns in data. Can't find values for the following variables: "
+                f" {set(self.nodes()) - set(data.columns)}"
+            )
+
+        # Check for missing values
+        if data.isnull().any().any():
+            raise ValueError(
+                "Data contains missing values. Log-likelihood computation "
+                "requires complete observations."
+            )
+
+        self.check_model()
+
+        # Create mapping from node to CPD for efficient lookup
+        node_to_cpd = {cpd.variable: cpd for cpd in self.get_cpds()}
```

Comment:
`node_to_cpd` is built from `self.get_cpds()` but `get_cpds()` could return empty and fail or raise KeyError. Given that users may call `log_likelihood` before fitting/adding cpds, error should be clear

Reply:
Ok, we can either validate the result of `self.get_cpds` or the dictionary `node_to_cpd`. For example, if `get_cpds()` returns empty, then `note_to_cpd` will be empty as well.

We can opt for something like:
```suggestion
        # Create mapping from node to CPD for efficient lookup
        node_to_cpd = {cpd.variable: cpd for cpd in self.get_cpds()}

        if not node_to_cpd:
            raise ValueError("No CPDs found in the model.")
```

To manage possible KeyError exceptions, we can define a `cpd` variable and wrap it with a `try-except` statement:
```python
    # Compute log-probability for each node in this time slice
    for node in nodes_t:
        try:
            cpd = node_to_cpd[node]
        except:
            raise KeyError(
                f"CPD for node {node} not found in the model."
            )

        node_log_prob = self._log_likelihood_node(
            node, data.iloc[i], cpd
        )
        sample_log_likelihood += node_log_prob
```
since `node_to_cpd[node]` could raise that kind of exception.